File: JavaProblems/MaximumProductSubarray.java
import java.util.Arrays;

// * dynamic programming
class MaximumProductSubarray {
    public int maxProduct(int[] nums) {
        int res = Arrays.stream(nums).max().getAsInt();  // 0 -> [-1]
        int curMin = 1, curMax = 1;

        for (int n : nums) {
            if (n == 0) {  // ignore 0 to not mess up the product
                curMin = 1;
                curMax = 1;
                continue;
            }

            int tmp = curMax * n;
            curMax = Math.max(n, Math.max(tmp, n * curMin));
            curMin = Math.min(n, Math.min(tmp, n * curMin));

            res = Math.max(res, curMax);
        }

        return res;
    }

    // Overall time complexity: O(n), space complexity: O(1)
    static void main() {
        var sol = new MaximumProductSubarray();
        System.out.println(sol.maxProduct(new int[]{2, 3, -2, 4}));
        System.out.println(sol.maxProduct(new int[]{-2, 0, -1}));
    }
}

File: PythonProblems/maximum_product_subarray.py
# * dynamic programming
class Solution:
    def maxProduct(self, nums):
        res = max(nums)  # 0 -> [-1]
        curMin, curMax = 1, 1

        for n in nums:
            if n == 0:  # ignore 0 to not mess up the product
                curMin, curMax = 1, 1
                continue

            tmp = curMax * n
            curMax = max(tmp, n * curMin, n)
            curMin = min(tmp, n * curMin, n)
            res = max(res, curMax, curMin)

        return res


# Overall time complexity: O(n), space complexity: O(1)
print(Solution().maxProduct([2, 3, -2, 4]))
print(Solution().maxProduct([-2, 0, -1]))
